cpp/src/text/normalize.cu

```diff
@@ -261,4 +255,366 @@ std::unique_ptr<cudf::column> normalize_characters(cudf::strings_column_view con
+{
+  if (input.is_empty()) { return cudf::make_empty_column(cudf::data_type{cudf::type_id::STRING}); }
+
+  auto const first_offset  = (input.offset() == 0) ? 0
+                                                   : cudf::strings::detail::get_offset_value(
+                                                      input.offsets(), input.offset(), stream);
+  auto const last_offset   = (input.offset() == 0 && input.size() == input.offsets().size() - 1)
+                               ? input.chars_size(stream)
+                               : cudf::strings::detail::get_offset_value(
+                                 input.offsets(), input.size() + input.offset(), stream);
+  auto const chars_size    = last_offset - first_offset;
+  auto const d_input_chars = input.chars_begin(stream) + first_offset;
+
+  if (chars_size == 0) { return std::make_unique<cudf::column>(input.parent(), stream, mr); }
+
+  constexpr int64_t block_size = 256;
+  cudf::detail::grid_1d grid{chars_size, block_size};
+  auto const max_new_char_total = cudf::util::round_up_safe(chars_size, block_size) * MAX_NEW_CHARS;
+
+  auto const parameters = normalizer._impl;
+
+  auto d_normalized = rmm::device_uvector<uint32_t>(max_new_char_total, stream);
+  data_normalizer_kernel<<<grid.num_blocks, grid.num_threads_per_block, 0, stream.value()>>>(
+    d_input_chars,
+    chars_size,
+    parameters->cp_metadata.data(),
+    parameters->aux_table.data(),
+    parameters->do_lower_case,
+    d_normalized.data());
+
+  // This removes space added around any special tokens in the form of [ttt].
+  // An alternate approach is to do a multi-replace of '[ ttt ]' with '[ttt]' right
+  // before returning the output strings column.
+  auto const special_tokens = parameters->get_special_tokens();
+  if (!special_tokens.empty()) {
+    special_tokens_kernel<<<grid.num_blocks, grid.num_threads_per_block, 0, stream.value()>>>(
+      d_normalized.data(), chars_size, special_tokens);
+  }
+
+  // Use segmented-reduce over the non-zero codepoints to get the size of the output rows
+  auto const input_offsets =
+    cudf::detail::offsetalator_factory::make_input_iterator(input.offsets(), input.offset());
+  auto output_sizes =
+    compute_sizes(d_normalized, input_offsets, first_offset, input.size(), stream);
+
+  // convert the sizes to offsets
+  auto [offsets, total_size] = cudf::strings::detail::make_offsets_child_column(
+    output_sizes.begin(), output_sizes.end(), stream, mr);
+
+  // create output chars by calling remove_copy(0) on the bytes in d_normalized
+  auto chars       = rmm::device_uvector<char>(total_size, stream, mr);
+  auto const begin = reinterpret_cast<char const*>(d_normalized.begin());
+  // the remove() above speeds up the remove_copy() by roughly 10%
+  auto const end =
+    reinterpret_cast<char const*>(remove_safe(d_normalized.begin(), d_normalized.end(), 0, stream));
+  remove_copy_safe(begin, end, chars.data(), 0, stream);
```

Comment:
if `remove_safe` already removed `0`, do we need `remove_copy_safe` again? why no just copy?

Reply:
The `remove_safe` actually removes all the `int32(0)` which is 4 bytes while `remove_copy_safe` here removes all the single byte `char(0)`. The 1st one removes a large number of 4-byte 0s while the 2nd one removes the rest. Technically we can just call `remove_copy_safe` (removes all 0-bytes) without the `remove_copy` but it turns out it is faster to call both in turn as coded here.

Reply:
Great thinking! This is a very clever approach.